src/sentry/grouping/component.py

```diff
@@ -319,3 +317,41 @@ class TemplateGroupingComponent(
     BaseGroupingComponent[ContextLineGroupingComponent | FilenameGroupingComponent]
 ):
     id: str = "template"
+
+
+# Wrapper components used to link component trees to variants
+
+
+class DefaultGroupingComponent(
+    BaseGroupingComponent[
+        CSPGroupingComponent
+        | ExpectCTGroupingComponent
+        | ExpectStapleGroupingComponent
+        | HPKPGroupingComponent
+        | MessageGroupingComponent
+        | TemplateGroupingComponent
+    ]
```

Comment:
this feels very very wrong.  a generic shouldn't need to enumerate its subclasses like this

and looking at BaseGroupingComponent it seems to be defined incorrectly (the `Any` there breaks all typechecking for instance) and the annotations on the class itself should be in `__init__`

#80722 probably should have had the typing group on it

Reply:
These are essentially typed nodes in a tree, and each type of node can only have other node types as children - that's what this is describing.